### PR TITLE
Stop thread spam

### PR DESCRIPTION
**Pull request by brianlovin**
**Status**
- [ ] WIP
- [x] Ready for review
- [ ] Needs testing

**Deploy after merge (delete what needn't be deployed)**
- api

**Release notes for users (delete if codebase-only change)**
- Adds protections to Spectrum to avoid people from spamming new threads across communities

Notes for @mxstbr:
- Prevents user from posting more than 3 threads within a 30m time window
- Prevents user from sending 3 threads in the same community within a 30m time window
- Runs a string similarity check on thread title and body with any threads posted in the previous 30m - if string similarity returns >80% for either title or body, the thread will not be published (until 30m window passes)
- After thread is published, we run a toxicity check - if both services agree the thread is >=98% toxic the thread *will* be published but *will not* send email notifications and will not add to the user's rep. It will still be search indexed in Vulcan.

Looking for feedback on constants here (30m timeframe, 3 thread limit, 98% toxicity confidence, 80% spam confidence)

### CODE REVIEW

**Comment by brianlovin** [comment]
Actually the toxicity works this way:
- If thread is agreed by both services to be >98% toxic it won't send both *in-app* and *email* notifications - does this feel right? Or should it send in-app notifications no matter what?

Skipping the entire notifications queue will also skip all the logic in `athena/queues/thread-notification` - which includes checking for @mentions. Do we care about that?

**Comment by mxstbr** [comment]
> Prevents user from posting more than 3 threads within a 30m time window
> Prevents user from sending 3 threads in the same community within a 30m time window

This seems very low and I assume will hit normal users fairly often. (or at least more often than we'd like) I'd heavily increase the number and decrease the timeframe to hone in on the spammers that only copy/paste or write garbage content, maybe something like 10 threads in 10 minutes? I'm not sure if e.g. 10/10 is better than 5/5 but you get my drift, anything above 1 thread per 1 minute seems excessive and we can assume that it's not legitimate content.

> Skipping the entire notifications queue will also skip all the logic in athena/queues/thread-notification - which includes checking for @mentions. Do we care about that?

That makes sense to me, `fuck you @brian` probably shouldn't show you that as a notification :laughing:

File: api/mutations/thread/publishThread.js [hunk 1]
```diff
@@ -52,33 +67,21 @@ export default async (
...
   ] = await Promise.all([
-    getCurrentUserChannelPermissions,
-    getCurrentUserCommunityPermissions,
-    getChannel,
-    getCommunity,
+    getUserPermissionsInChannel(thread.channelId, currentUser.id),
+    getUserPermissionsInCommunity(thread.communityId, currentUser.id),
+    getChannelById(thread.channelId),
+    getCommunityById(thread.communityId),
+    getThreadsByUserAsSpamCheck(currentUser.id),
```
Comment (mxstbr) [inline api/mutations/thread/publishThread.js]: These should all be loaders. We already have the first 4, only the new one doesn't have a loader.

File: api/mutations/thread/publishThread.js [hunk 2]
```diff
@@ -162,14 +226,67 @@ export default async (
...
+      ? threadBodyToPlainText(thread.content.body)
+      : '';
+    const title = thread.content.title;
+    const text = `${title} ${body}`;
+
+    const scores = await Promise.all([
+      getSpectrumScore(text, dbThread.id, dbThread.creatorId),
+      getPerspectiveScore(text),
+    ]).catch(err =>
+      console.error(
```
Comment (mxstbr) [inline api/mutations/thread/publishThread.js]: We should make sure to properly handle the error case here b/c this actually happens fairly often if you look at the athena logs (~once per hour)

**Review by mxstbr — COMMENTED** [review]
Overall, this seems like an amazing addition already, pending my comments

**Comment by mxstbr** [comment]
> I'm not sure if e.g. 10/10 is better than 5/5 

I actually think 5/5 is getting somewhere nice—no legitimate user can create six threads full of beautiful content in five minutes, buuuut if they do manage to make it happen it's not too much to ask to wait 2 minutes or whatever to post the sixth one.

It also means spammers won't be able to post 10 threads in a row, but only five.

**Comment by mxstbr** [comment]
Maybe it should be 3/5 if we want to be more strict, I don't even think anybody posting three threads in five minutes has good intentions/is producing good content. Butttt again if they are it's not too much to ask to wait 2 minutes to post a fourth thread.

File: api/mutations/thread/publishThread.js (hunk 2 above)
Comment (brianlovin) [reply api/mutations/thread/publishThread.js]: It's just random Perspective failures, no idea why - catching the error here should still keep the overall mutation safe right?

**Comment by brianlovin** [comment]
Talked offline, we're moving to a 3/10 constraint for spam.

File: api/mutations/thread/publishThread.js (hunk 1 above)
Comment (brianlovin) [reply api/mutations/thread/publishThread.js]: Not totally sure why we'd want a loader for the last one, or even the value of loaders in mutations (they have way more probability of causing cache mismatches than of providing meaningful mutation speed, imo). What do you think?

**Comment by brianlovin** [comment]
Pushed some more fixes and set up an email queue to alert admins when we get a spam report so we can fine tune our algorithms around string-matching sensitivity and post frequency limits.

File: api/mutations/thread/publishThread.js [hunk 3]
```diff
@@ -102,27 +105,104 @@ export default async (
...
+  // if user has published other threads in the last hour, check for spam
+  if (
+    usersPreviousPublishedThreads &&
+    usersPreviousPublishedThreads.length > 0
+  ) {
+    debug(
+      'User has posted at least once in the previous 10m - running spam checks'
     );
+
+    if (usersPreviousPublishedThreads.length >= 3 && !isOwnerOrModerator) {
```
Comment (mxstbr) [inline api/mutations/thread/publishThread.js]: I wouldn't be 100% lenient with owners/mods. Imagine if the owner account of the React community got hacked and started pushing dozens of posts. Let's still be fairly strict and do something like 5/10 for owners/mods?

File: api/mutations/thread/publishThread.js [hunk 4]
```diff
@@ -102,27 +105,104 @@ export default async (
...
+        debug(`Body similarity score for spam check: ${bodySimilarity}`);
+        if (bodySimilarity > 0.8) return true;
+      }
+
+      return false;
+    });
+
+    const isSpamming = checkForSpam.filter(Boolean).length > 0;
+
+    if (isSpamming && !isOwnerOrModerator) {
```
Comment (mxstbr) [inline api/mutations/thread/publishThread.js]: Again we probably shouldn't be 100% lenient for owners/mods either

File: api/mutations/thread/publishThread.js [hunk 5]
```diff
@@ -162,14 +242,77 @@ export default async (
...
-  const dbThread = await publishThread(threadObject, currentUser.id);
+  // $FlowFixMe
+  const dbThread: DBThread = await publishThread(threadObject, currentUser.id);
+
+  // we check for toxicity here only to determine whether or not to send
+  // email notifications - the thread will be published regardless, but we can
+  // prevent some abuse and spam if we ensure people dont get email notifications
+  // with titles like "fuck you"
+  const checkToxicity = async () => {
+    if (isOwnerOrModerator) return false;
```
Comment (mxstbr) [inline api/mutations/thread/publishThread.js]: I wouldn't be lenient at all here, doesn't matter who publishes a toxic thread nobody should get a notification for that.

File: api/mutations/thread/publishThread.js [hunk 6]
```diff
@@ -162,14 +242,77 @@ export default async (
...
+      : '';
+    const title = thread.content.title;
+    const text = `${title} ${body}`;
+
+    // $FlowFixMe
+    const scores = await Promise.all([
+      // $FlowFixMe
+      getSpectrumScore(text, dbThread.id, dbThread.creatorId),
+      // $FlowFixMe
+      getPerspectiveScore(text),
```
Comment (mxstbr) [inline api/mutations/thread/publishThread.js]: Why did you have to FlowFixMe here? I think we should maybe catch those two promise errors individually and set their score to 0:

```JS
const scores = await Promise.all([
  getSpectrumScore(text, dbThread.id, dbThread.creatorId).catch(err => 0),
  getPerspectiveScore(text).catch(err => 0),
]).catch(err => console.error(bla bla))
```

File: api/mutations/thread/publishThread.js [hunk 7]
```diff
@@ -162,14 +242,77 @@ export default async (
...
+    const perspectiveScore = scores && scores[1];
+
+    // if neither models returned results
+    if (!spectrumScore && !perspectiveScore) {
+      debug('Toxicity checks from providers say not toxic');
+      return false;
+    }
+
+    // if both services agree that the thread is >= 98% toxic
+    if ((spectrumScore + perspectiveScore) / 2 >= 0.98) {
```
Comment (mxstbr) [inline api/mutations/thread/publishThread.js]: I just checked my emails and I cannot find a single instance so far where the toxicity score across both providers is > 98%. Let's move this down a notch to 90%, remove the emails we're currently getting and only send those emails when this hits?

File: api/mutations/thread/publishThread.js [hunk 8]
```diff
@@ -162,14 +242,77 @@ export default async (
...
+    debug(
+      'Thread determined to be toxic, not sending notifications or adding rep'
+    );
+    // generate an alert for admins
+    _adminProcessToxicThreadQueue.add({ thread: dbThread });
+  } else {
+    debug('Thread is not toxic, send notifications and add rep');
+    // thread is clean, send notifications and process reputation
+    sendThreadNotificationQueue.add({ thread: dbThread });
+    processReputationEventQueue.add({
```
Comment (mxstbr) [inline api/mutations/thread/publishThread.js:312]: We should still process rep for threads that say "fuck you", who knows where people might be posting that and what their norms for communication are.

**Comment by brianlovin** [comment]
Agreed on your feedback, thanks for pushing here! Pushed fixes for all those.

**Review by mxstbr — DISMISSED** [review]
This looks fine to me, let's wait for @uberbryn's thoughts then ship it!

**Review by mxstbr — APPROVED** [review]
LGTM

**Comment by superbryntendo** [comment]
My only concern is for communities that have things like bug reports/feature requests channels where it's likely that many people will post several threads in a row. With the 80% similarity checks, especially, I'd be surprised if we needed restrictive limits.

That said, testing will answer this for us. Let's ship it.

**Comment by superbryntendo** [comment]
oh, also there are examples like Sarah posting several podcast episodes in a row. No shared content, but several threads in a short timeframe.

**Comment by brianlovin** [comment]
@uberbryn it will be a 10 min timeframe; do you know how many threads she posts in that range? Feels like a pretty good edge case, but you're right testing will definitely help us find the right boundaries here.

**Pull request merged**